Fix ASM ambiguity

willcl-ark opened the pull request:
Closes: #27795
Closes: #7996

Previously `ScriptToAsmStr` returned hex-encoded values, except if data length was <= 4 bytes, in which case it displayed using decimal encoding.

Fix the ASM representation by specifying an unambiguous encoding:

- Drop `OP_` prefix from all opcodes
- For non-minimal pushes prefix with the opcode and enclose pushed hex value in angle brackets `PUSHDATA1<0100>`
- For minimal pushes:
  - If > 5 bytes display pushed hex value enclosed in angle brackets `<...>`
  - If \<= 5 bytes:
    - If minimally-encoded display pushed value in decimal without angle brackets `42`
    - If not minimally-encoded display pushed hex value enclosed in angle brackets `<4200>`
  - Display undecodable bytes using `UNDECODABLE(...)`
  
This should permit unambiguous decoding in the future.

willcl-ark committed:
doc: release note for asm repr script changes

ajtowns commented on test/functional/rpc_decodescript.py:
```diff
@@ -133,7 +133,7 @@ def decodescript_script_pub_key(self):
         rpc_result = self.nodes[0].decodescript(cltv_script)
         assert_equal('nonstandard', rpc_result['type'])
-        assert_equal('OP_IF ' + public_key + ' OP_CHECKSIGVERIFY OP_ELSE 500000 OP_CHECKLOCKTIMEVERIFY OP_DROP OP_ENDIF ' + public_key + ' OP_CHECKSIG', rpc_result['asm'])
+        assert_equal('OP_IF ' + public_key + ' OP_CHECKSIGVERIFY OP_ELSE 20a107 OP_CHECKLOCKTIMEVERIFY OP_DROP OP_ENDIF ' + public_key + ' OP_CHECKSIG', rpc_result['asm'])
```
This seems much harder to understand to me -- you need to convert it to `0x07a120` then convert that to decimal to get back to the nice round number block height? In particular mistaking it for `0x20a107` (2,138,375) would be particularly misleading.

ajtowns commented on test/functional/wallet_send.py:
```diff
@@ -170,7 +170,7 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
                     assert_greater_than(from_balance_before - from_balance, amount)
             else:
-                assert next((out for out in tx["vout"] if out["scriptPubKey"]["asm"] == "OP_RETURN 35"), None)
+                assert next((out for out in tx["vout"] if out["scriptPubKey"]["asm"] == "OP_RETURN 23"), None)
```
This also seems confusing, particularly if it appears in something like `DUP SIZE 20 EQUALVERIFY HASH160 <x> EQUAL` -- whoops, that's actually saying the input should be 32 bytes, not 20 bytes.

MatthewLM commented:
In the dart library coinlib (https://pub.dev/packages/coinlib) I followed the same approach of encoding everything in HEX with integers encoded as LE. However, 0 and -1 are encoded as 0 and -1. See the code here: https://github.com/peercoin/coinlib/blob/master/coinlib/lib/src/scripts/operations.dart

I did this to avoid changing the ASM too much. However, I'm happy to move towards using `0x` for hex and presenting decimal for smaller data of up-to 32bits if this is introduced in Bitcoin. This would make the ASM easy to understand and read.

willcl-ark commented:
@ajtowns @MatthewLM thank for your comments.

I agree that after these changes some ASM fields will be less human-readable-friendly in LE hex.

Prefixing with `0d` was another possibility discussed previously, but there still exists a chance that someone will interpret these as hex-encoded values, as `d` is a valid hex char (whereas `x` is clearly not decimal or hex). e.g this is not necessarily _that_ much clearer, but does offer some improvement:

```sh
$ bitcoin-cli -regtest decodescript 0511121314150457c74942 | jq .asm
"1112131415 0d1112131415"

$ src/bitcoin-cli -regtest decodescript 0512345678900320A107b1 | jq .asm
"1234567890 0d500000 OP_CHECKLOCKTIMEVERIFY"

```

Thinking about this more, it feels like the question to be answered is perhaps "is the `asm` field supposed to be human or machine readable?" If the former, then maintaining decimal for small values, and adding a hex-prefix for > 4 byte values might be most desirable, even it it leaves some ambiguity in some circumstances?

```sh
$ src/bitcoin-cli -regtest decodescript 0512345678900320A107b1 | jq .asm
"0x1234567890 500000 OP_CHECKLOCKTIMEVERIFY"
```

This would create a moderately inconvenient interface in a few of the tests (and possibly some downstream users relying on this output) where hex prefixes, if present, have to be stripped from rpc output, but perhaps that's not too terrible.

If we are designing the asm output to be a canonical, machine-readable format then my opinion remains that using hex-only makes the most sense.

FWIW [btcdeb](https://github.com/bitcoin-core/btcdeb) interprets all values as decimal but prints a warning that they are ambiguous:

```sh
$ btcc 1234567890 500000 OP_CHECKLOCKTIMEVERIFY
warning: ambiguous input 1234567890 is interpreted as a numeric value; use 0x1234567890 to force into hexadecimal interpretation
warning: ambiguous input 500000 is interpreted as a numeric value; use 0x500000 to force into hexadecimal interpretation
04d20296490320a107b1

$ btcc 0x1234567890 500000 OP_CHECKLOCKTIMEVERIFY
warning: ambiguous input 500000 is interpreted as a numeric value; use 0x500000 to force into hexadecimal interpretation
0512345678900320a107b1

```

And consequently generates a different script without manually specifying hex. I don't know what other tools do by default.

MatthewLM commented:
I agree that `0d` is not a good idea, much better to go with `0x` for hex.

ajtowns commented:
> Thinking about this more, it feels like the question to be answered is perhaps "is the `asm` field supposed to be human or machine readable?"

I think it's human readable? For machines we could just give the raw hex script and not bother with an asm format.

Anyway, how about going for something more left field and having `decodescript 0511121314150457c74942` output `[1112131415] 1112131415` where the square brackets indicated "quoted hexidecimal", and values without square brackets are always decimal numbers within `CScriptNum` range? That would still be pretty clear for pubkeys:

```
2 [027d28a3580d0823dbee34f1df927ee58664eb8349e79b94017ddaf3f7e159153f] [03a3af0f49a21d29106ebeef9b3c3d69fc375c856a7153d97156a5b7a161ca6c25] 2 OP_CHECKMULTISIG
```

Could perhaps also represent non-minimal pushes that way, eg the different encodings of -1 or `[ff]`, something like:

```
 decodescript 4f -> -1
 decodescript 0181 -> [0:81]
 decodescript 4c0181 -> [1:81]
 decodescript 4d010081 -> [2:81]
 decodescript 4e0100000081 -> [4:81]
```

willcl-ark committed:
test: add asm repr asm_hex helper fn

A simple helper function to angle bracket a hex string.

willcl-ark commented:
> I think it's human readable? For machines we could just give the raw hex script and not bother with an asm format.

This has always been my interpretation too.

> Anyway, how about going for something more left field...

Square brackets is something I could get behind; to me it is clear from your example that the brackets are indicating _something_, and just by looking at them I can see in most cases that they're hex values. To use the previous (less common?) example formatted with the brackets:

```sh
$ bitcoin-cli -regtest decodescript 0511121314150457c74942 | jq .asm
"[1112131415] 1112131415"
```

IMO it's again clear to me now that there's _something_ different about these two values, and if I didn't know what it was then I should find out.

One question I have is whether this will interfere with the scripthash types too badly? e.g. Is this nested bracket undesirable?

```sh
$ src/bitcoin-cli -regtest decoderawtransaction 0100000001696a20784a2c70143f634e95227dbdfdf0ecd51647052e70854512235f5986ca010000008a47304402207174775824bec6c2700023309a168231ec80b82c6069282f5133e6f11cbb04460220570edc55c7c5da2ca687ebd0372d3546ebc3f810516a002350cac72dfe192dfb014104d3f898e6487787910a690410b7a917ef198905c27fb9d3b0a42da12aceae0544fc7088d239d9a48f2828a15a09e84043001f27cc80d162cb95404e1210161536ffffffff0100e1f505000000001976a914eb6c6e0cdb2d256a32d97b8df1fc75d1920d9bca88ac00000000 | jq .vin[].scriptSig.asm
"[304402207174775824bec6c2700023309a168231ec80b82c6069282f5133e6f11cbb04460220570edc55c7c5da2ca687ebd0372d3546ebc3f810516a002350cac72dfe192dfb[ALL]] [04d3f898e6487787910a690410b7a917ef198905c27fb9d3b0a42da12aceae0544fc7088d239d9a48f2828a15a09e84043001f27cc80d162cb95404e1210161536]"

```

specifically:

```sh
[... 2dfb[ALL]]
```

It might be preferable to have this format instead?

```sh
[... 2dfb][ALL]
```

MatthewLM commented:
`0x` is much more standard and recognisable than `[]`.

ajtowns commented:
> `0x` is much more standard and recognisable than `[]`.

It's pretty misleading though, since `CScriptNum` is little-endian: if you see `0x0100` you would expect that to be 256, but for us it would actually be non-minimally-encoded 1; For us, `0x0201` would actually be less than `0x0102`, etc.

Note this applies to block hashes, as well, so if you provide block 816651's header and execute HASH256, you end up with `[7d6368795d9675a198bd1dbe3d87dbdbbc3e0f3560a202000000000000000000]`; if you were to write `0x00000000000000000002a260350f3ebcdbdb873dbe1dbd98a175965d7968637d` to refer to the block hash, the same way we do in kernel/chainparams.cpp, you'd end up with the wrong thing. But you can't have "0x" just reverse things either, as that would mess up 33-byte pubkeys (they'd now end with 02 or 03 instead of beginning with it) and signatures.

ajtowns commented:
> One question I have is whether this will interfere with the scripthash types too badly? e.g. Is this nested bracket undesirable?
> 
> specifically:
> 
> ```shell
> [... 2dfb[ALL]]
> ```
> 
> It might be preferable to have this format instead?
> 
> ```shell
> [... 2dfb][ALL]
> ```

I guess it being inside the brackets seems slightly clearer to me? Could have different brackets for the two things, eg `[... 2dfb<ALL>]`. Could perhaps also drop the feature entirely (reverting #5264); note that we don't display sighash flags for taproot signatures.

MatthewLM commented:
> It's pretty misleading though, since `CScriptNum` is little-endian: if you see `0x0100` you would expect that to be 256, but for us it would actually be non-minimally-encoded 1; For us, `0x0201` would actually be less than `0x0102`, etc.

Those smaller numbers will be shown in decimal still. For larger pushdata, hex is already shown in the little-endian encoded byte order. If `0x` is too suggestive of a big-endian hex number (a fair point), then maybe `bytes(123456789abcdef)` is clearer and less ambiguous (maybe use `push` instead of `bytes`). If following that scheme, signatures could be something like `sig(...2dfb, ALL)`. Possibly public keys could be `pk(...)`.

ajtowns commented:
> My contribution to this bikeshedding would be:
>     * `OP_n` -> `n` (so `-1`, `0`, `1`, ..., `10`, ..., `16`).
>     * Drop "OP_" prefix for other opcodes too; they're redundant.
>     * Drop the sighash byte type decoding. I don't feel it matches what asm is trying to do.

:+1:

> * For non-minimal pushes, use `PUSHDATA1<...>` or `PUSHDATA2<...>` (with the contents of `...` using the same rules as for minimal pushes).

Also `PUSHDATA4<...>` for completeness, presumably. A direct push of values -1 and 1..16 is also non-minimal (ie, `0101` instead of `51`), could perhaps use `PUSHDATA<..>` (without 1, 2 or 4) for that case.

> * Other _minimal_ pushes (so direct pushes, as well as OP_PUSHDATA1 over 75 bytes, and OP_PUSHDATA2 over 255 bytes):       
> * If the data pushed is a minimally-encoded positive number between 0 and 2^39-1 (`OP_CLTV` / `OP_CSV` accept 5-byte numbers) -> `<0x12345678>` (BE hex, as that's what a human expects for a numeric value with 0x).

Provided you've already dealt with non-minimally encoded -1 and 1..16, using decimals seems nicer here? You can cope with negatives cleanly (`-2` vs `<82>`), and it's one less encoding variation?

> * Optionally, for P2SH(-like) scriptSigs, the final redeemScript push could be recursively decoded (so inside `<...>`, put the asm decoding of the script being pushed), while remaining fully unambiguous.

That would be nice. Probably requires remembering info from the scriptPubKey and using it when decoding the scriptSig or witness.

willcl-ark committed:
script: add GetOpNameAsm helper fn

Strip unneccesary leading "OP_" prefix from opcodes when they are
rendered as ASM.

willcl-ark committed:
script: don't decode sighash flags in ASM repr

We don't decode these on taproot signatures already, modify non-taproot
signatures to match.

MatthewLM commented:
> Provided you've already dealt with non-minimally encoded -1 and 1..16, using decimals seems nicer here? You can cope with negatives cleanly (`-2` vs `<82>`), and it's one less encoding variation?

I was thinking it would be nice to retain decimal but, due to ambiguity that could arise with CLTV and CSV, I'm not sure.

Having embedded ASM for redeem scripts and Tapscripts would be very useful.

willcl-ark committed:
script: drop OP_ prefix from ASM repr

willcl-ark committed:
script: remove ambiguity from ASM repr

* Return minimal pushes < 5 bytes as decimal
* Return non-minimal pushes < 5 bytes wrapped as OP_CODE<...>
* Return pushes > 5 bytes as hex wrapped as <...>

ajtowns commented:
> I was thinking it would be nice to retain decimal but, due to ambiguity that could arise with CLTV and CSV, I'm not sure.

What ambiguity?

MatthewLM commented:
For 32-bit signed integers, the most significant bit is negative, but that's not true for CLTV and CSV which uses 5 bytes. So a push data could represent a negative 32-bit signed integer, or a positive 40-bit signed integer if I understand the `CScriptNum` without looking more closely at it.

MatthewLM commented:
@sipa I'll have to look at the code. I image it is two's complement? Then the negative bit will be different for each type of integer meaning that a 4 byte pushdata that would encode a negative 32-bit signed integer could also be interpreted as a positive 40-bit signed integer.

MatthewLM commented:
@sipa Sorry, you are right. I looked at the code again. Bitcoin doesn't use two's complement for this (I have to keep in mind that Bitcoin has a lot of quirks). It uses a sign bit that negates the other bits, and this sign bit depends on the size of the push data (whatever the last byte is), so there is no ambiguity.

Therefore I think minimally pushed numbers ought to be displayed as decimal for readability reasons as you now suggest.

> * Inside `<...>` instead of having hex data, it's also allowed to put another decoded script (as long the script doesn't consist of exactly one 2-digit number push, because that'd be ambiguous...).

Or any other decimal that has an even number of digits. Maybe use something like `SCRIPT<...>`?

willcl-ark commented:
I like where this is going now. Thanks for the conversation.

I've written a branch implementing this [previous version](https://github.com/bitcoin/bitcoin/pull/28824#issuecomment-1810204432) as I wanted to see what the changes would look like (although I'm not convinced I'm handling `OP_1NEGATE` properly yet, even though tests are passing...)

> Any minimally-encoded number, using a minimal push (OP_n, or direct push for numbers that cannot be encoded using OP_n) of an integer between 231 and 239-1, is just encoded in decimal directly (no <> or anything).

I think this will make the representation even nicer to interpret and will gladly make this change.

> Inside <...> instead of having hex data, it's also allowed to put another decoded script (as long the script doesn't consist of exactly one 2-digit number push, because that'd be ambiguous...).

I'll take a stab at implementing these nested scripts as well, as I think they'd be a nice improvement too...

ajtowns commented:
>  * Any minimally-encoded number, using a minimal push (`OP_n`, or direct push for numbers that cannot be encoded using OP_n) of an integer between _231_ and _239-1_, is just encoded in decimal directly (no `<>` or anything).

`-2**31` and `2**39-1` presumably? (original lacks the negative sign, and github can't quote superscript correctly apparently)
 
> * Any other minimal pushes (so direct push <= 75 bytes, PUSHDATA1 > 75 bytes, PUSHDATA2 > 255 bytes) become `<...>` with ... the hex-encoding (in wire byte order) of the push.

Per `script/script.cpp:CheckMinimalPush()`, `0101` isn't a minimal push, so I think it'd be better to drop the "minimal push" phrasing and just use the definition currently in brackets. (also; PUSHDATA4 > 65535 bytes)

> * Inside `<...>` instead of having hex data, it's also allowed to put another decoded script (as long the script doesn't consist of exactly one 2-digit number push, because that'd be ambiguous...).

Any push of a single positive number between 10-99; 1,000-9,999; 100,000-999,999; 10,000,000-99,999,999; 1,000,000,000-9,999,999,999; and 100,000,000,000-549,755,813,887 would be ambiguous, no?

MatthewLM commented:
Could we use `UNKNOWN_n` where n denotes the op code hex/decimal such as `UNKOWN_FE` for `0xfe`?

willcl-ark commented:
Marked as draft for now and pushed what I have worked on so far.

I haven't looked at implementing undecodable bytes or nested scripts yet.

luke-jr commented on src/core_write.cpp:
```diff
@@ -109,33 +115,46 @@ std::string ScriptToAsmStr(const CScript& script, const bool fAttemptSighashDeco
+                    str += strprintf("%lld", n.GetInt64());
                 } else {
-                    str += HexStr(vch);
+                    // Otherwise display the push as LE hex enclosed in angle brackets
```
It's not little endian, I agree.

willcl-ark commented on src/core_write.cpp:
```diff
@@ -109,33 +115,46 @@ std::string ScriptToAsmStr(const CScript& script, const bool fAttemptSighashDeco
+            if (CheckMinimalPush(vch, opcode)) {
+                if (vch.size() <= 5 ) {
+                    // Return decimal for minimially-encoded values <= 5 bytes (OP_CLTV / OP_CSV accept 5-byte numbers)
+                    CScriptNum n(vch, false, 5);
```
Agree that it should be `true` as a belt-and-suspenders, but we are inside `if (CheckMinimalPush(vch, opcode))` already here, so shouldn't happen?

ajtowns replied:
Byte code `024000` is a minimal push of the two-byte vector `4000` (so passes `CheckMinimalPush`) but it's not a minimal `CScriptNum` since it just evaluates to 64, and should have been `0140` (so fails `fRequireMinimal` due to the trailing 0).

willcl-ark replied:
Thanks for the explanations.

I've refactored out handling pushed data, as I wasn't enjoying all the nested if statements, and notice that we are currently handling "non-minimal pushes" the same as minimal pushes with non-minimally-encoded values, which is to use `OPCODE<hex>`. This seems ok, and it will not be a lossy decode, but do we want to differentiate between the two pathways to this decode format?

See first and last case in snippet to illustrate what I mean above:

```cpp
/*
 * Format pushed bytes into appropriate ASM format
 */
std::string FormatPushDataAsm(const std::vector<unsigned char>& vch, opcodetype opcode)
{
    // Use OPCODE<hex> for non-minimal pushes
    // TODO: There are no tests for this currently
    if (!CheckMinimalPush(vch, opcode)) {
        return GetOpNameAsm(opcode) + BracketStr(HexStr(vch));
    }

    // Use <hex> for minimal pushes > 5 bytes
    if (!(vch.size() <= 5) ) {
        return BracketStr(HexStr(vch));
    }

    // Use decimal for minimally-encoded, minimal pushes <= 5 bytes
    // Note: OP_CLTV / OP_CSV accept 5-byte numbers
    try {
        CScriptNum n{vch, true, 5};
        return strprintf("%lld", n.GetInt64());
    }

    // Use OPCODE<hex> for non-minimally-encoded minimal pushes
    // TODO: There are no tests for this currently
    catch (scriptnum_error& e) {
        return GetOpNameAsm(opcode) + BracketStr(HexStr(vch));
    }
}
```

willcl-ark replied:
Sure, but what I mean is that in my current impl there are _two_ ways we can end up with `OPCODE<...>`:

1) Wasn't a minimal push
2) Was a minimal push <5 bytes, but wasn't a minimally-encoded decimal value

ajtowns replied:
Yeah. This just means changing the final `catch` to `return BracketStr(HexStr(vch));`, no?

ajtowns replied:
Also,

```c++
if (!(vch.size() <= 5) ) {
```

Can I introduce you to `>` ? :smile:

willcl-ark commented:
> Brainstormy idea for undecodable scripts.

Currently failure to fully parse in `GetScriptOp` will only return `false` if we run out of bytes to read either the length, or from the operand, and we don't store the `opcode` or the available remaining bytes from the iterator in `opcodeRet` or `pvchRet`. But we could change `GetScriptOpt` to store them the failure case too.

> `<...+N>` or `OPCODE<...+N>`

I don't think it makes much difference which flavour is chosen here; both would require the `N` to be parsed to calculate the total length and therefore determine the opcode that was used before it was encoded. As there are bytes missing, I don't think we can tell if this was a minimal push or not either. I'd probably slightly lean towards using `OPCODE<...+N>` as it's slightly more human-friendly?

willcl-ark committed:
script: return undecodable bytes

ajtowns commented:
> Brainstormy idea for undecodable scripts.
> 
> These are necessarily pushes (direct or OP_PUSHDATA[124]) whose prescribed length exceeds the end of the script. What about using `<...+N>` or `OPCODE<...+N>`, where `...` is hexadecimal as before, and `N` is a decimal number indicating how many bytes are missing?
> 
> EDIT: for `OP_PUSHDATA[124]` it's also possible that the 1-4 byte length field itself straddles the script end, and that isn't covered by this suggestion.

If you're adding extra encoding that need to be parsed (`+N`), maybe just have `UNPARSABLE<hex>` or `GARBAGE<hex>` to capture the trailing data that wasn't a full push? Seems less likely to be confusing (easy to miss a +30 at the end of a hex string), and works even if we change an OP_SUCCESS into a multibyte opcode in future?

ajtowns commented:
> Yeah, my only (rather minor) reservation about `UNPARSEABLE<...>` is that I think of `<...>` as a way of indicating "stuff being pushed",

Yeah, I guess being able to scan and say `<...>` means data is going onto the stack makes sense.

> So maybe just `UNPARSEABLE(...)`?

Could do `UNPARSEABLE:04010203` or something (colon instead of brackets). 

(Arguably, everything after the first OP_SUCCESS in a tapscript could also be considered trailing hex that shouldn't be decoded. Probably more useful to decode it anyway, though)
